sink/sink

```diff
@@ -303,6 +303,31 @@ class Buffer(object):
         result += os.read(self.fd, 1024)
         return result
 
+# Create the directory in a race free way
+def mkdir_and_chdir(base, identifier):
+    directory = os.path.abspath(os.path.join(base, identifier))
+
+    # 1. Create a temporary non-empty directory and get its handle
+    tempdir = tempfile.mkdtemp(prefix=identifier, dir=base)
+    os.close(os.open(os.path.join(tempdir, ".sink"), os.O_WRONLY | os.O_CREAT))
+    dirfd = os.open(tempdir, os.O_RDONLY)
+
+    # 2. Rename directory into place
+    while True:
+        try:
+            os.rename(tempdir, directory)
+            break
+        except OSError as ex:
+            # 3. If that raced, then move the target into our directory
+            if ex.errno == errno.EEXIST or ex.errno == errno.ENOTEMPTY:
+                os.rename(directory, tempfile.mkdtemp(prefix="backup.", dir="."))
```

Comment:
This places the backup directory into ".", which is normally the home directory of the user that runs the sink, not the newly created directory.